packages/cli/src/cmds/manifest/manifestHandler.ts

```diff
@@ -35,246 +11,45 @@ const ManifestSortOrder: Record<keyof SnapManifest, number> = {
+export async function manifestHandler({ writeManifest }: YargsArgs) {
   try {
-    ({ manifest } = validateNpmSnap(snapFiles, errorPrefix));
-  } catch (error) {
-    if (writeManifest && error instanceof ProgrammaticallyFixableSnapError) {
-      // If we get here, the files at least have the correct shape.
-      const partiallyValidatedFiles = snapFiles as SnapFiles;
-
-      let isInvalid = true;
-      const maxAttempts = Object.keys(SnapValidationFailureReason).length;
-
-      // Attempt to fix all fixable validation failure reasons. All such reasons
-      // are enumerated by the SnapValidationFailureReason enum, so we only
-      for (let attempts = 1; isInvalid && attempts <= maxAttempts; attempts++) {
-        manifest = fixManifest(partiallyValidatedFiles, error);
-
-        try {
-          validateNpmSnapManifest(
-            { ...partiallyValidatedFiles, manifest },
-            errorPrefix,
-          );
-
-          isInvalid = false;
-        } catch (nextValidationError) {
-          /* istanbul ignore next: this should be impossible */
-          if (
-            !(
-              nextValidationError instanceof ProgrammaticallyFixableSnapError
-            ) ||
-            (attempts === maxAttempts && !isInvalid)
-          ) {
-            throw new Error(
-              `Internal Error: Failed to fix manifest. This is a bug, please report it. Reason:\n${error.message}`,
-            );
-          }
-        }
-      }
-
-      didUpdate = true;
-    } else {
-      throw error;
-    }
-  }
-
-  // TypeScript doesn't see that the 'manifest' variable must be of type
-  // SnapManifest at this point, so we cast it.
-  const validatedManifest = manifest as SnapManifest;
-
-  // Check presence of recommended keys
-  const recommendedFields = ['repository'] as const;
-
-  const missingRecommendedFields = recommendedFields.filter(
-    (key) => !validatedManifest[key],
-  );
-
-  if (missingRecommendedFields.length > 0) {
-    logManifestWarning(
-      `Missing recommended package.json properties:\n${missingRecommendedFields.reduce(
-        (allMissing, currentField) => {
-          return `${allMissing}\t${currentField}\n`;
-        },
-        '',
-      )}`,
+    const { warnings, errors } = await checkManifest(
+      process.cwd(),
+      Boolean(writeManifest),
     );
-  }
-
-  // Validation complete, finish work and notify user.
 
-  if (writeManifest) {
-    try {
-      await fs.writeFile(
-        NpmSnapFileNames.Manifest,
-        `${JSON.stringify(getWritableManifest(validatedManifest), null, 2)}\n`,
-      );
-
-      if (didUpdate) {
-        console.log(`Manifest: Updated snap.manifest.json`);
-      }
-    } catch (error) {
-      throw new Error(
-        `${errorPrefix}Failed to update snap.manifest.json: ${error.message}`,
-      );
+    if (!writeManifest && errors.length > 0) {
+      console.error(`${ERROR_PREFIX}The manifest is invalid.`);
+      errors.forEach(logManifestError);
```

Comment:
Should we still throw to terminate with a non-zero exit code? We could just throw `The manifest is invalid, see errors above.` or something?

Reply:
Good point!